### ISSUE TITLE
Use parallel streams when uploading large files.

### ISSUE THREAD

**Issue by coryan** [googleapis/google-cloud-cpp#2951]
When uploading large files to GCS one can get better performance by running multiple uploads in parallel.  The file is divided in "parts", uploaded to a temporary object, and then the parts are reassembled using `Client::Compose`.  Then the parts are deleted using `Client::DeleteObject()` (with an specific version!).

We need to think about the names for these parts, maybe something like `<object-name-sha256>-tmp-parts/partNN`.

We should consider how these parts get deleted (object lifecycle rules?)

We should consider how we save the state of these uploads so the application can resume them. Though I presume "not resumable" is an option that would work for files that are not excessively large.

In this context "large" file is of course relative, but anything below 20MiB probably makes not sense to split.  And something like 1GiB is probably large enough that it deserves to be split.

Note that GCS cannot compose more than 1024 parts, and a single compose operation cannot compose more than 32 parts (so composing 32 parts each composed by 32 parts is the absolute limit).

Since the rest of the library is single threaded, the obvious implementation needs to create N threads to do the work.  Later we could implement a primitive that "uploads N things" at the same time using the Curl primitives for simultaneous transfers.

Also note that more than each upload uses 50MiB/s (that is about 400Mbps) and there are only 2Gpbs per core, so more than 5 uploads per core is not going to help.

**Comment by dopiera** [comment]
# Requirements
Some additional (nice-to-have) requirements I thought of:
* allow for separate processes writing to streams
* a process restart/crash shouldn't imply starting from scratch

# API
For the first step, I propose the following:
```C++

// Additional options:
// Let the user chose how many uploads stream to generate
struct NumShards {};
// Indicate the upload session id to restart
struct RestoreParallelUploadSession {};

struct ParallelUploadState {
  // Obtain a session id to be used with `RestoreParallelUploadSession`
  std::string resumable_session_id() const;
  // Return the number of shards of this upload session.
  std::size_t NumShards() const;
  // Get the writer for a shard of the upload.
  StatusOr<GcsObjectWriteStream> Shard(std::size_t idx);
  // Poll for the operation to complete.
  template< class Rep, class Period >
  StatusOr<ObjectMetadata> PollForCompletion(
      const std::chrono::duration<Rep, Period>& rel_time);
  // Cleanup all the temporary files
  Status Cleanup();
};

template <typename... Options>
StatusOr<ParallelUploadState> PrepareParallelUpload(
    Client client, std::string const &bucket, std::string const &object_name,
    std::size_t num_shards, std::string const &prefix, Options&&... options);
```
# Intended use:
```C++
  auto prefix = CreateRandomPrefix(...);
  if (!prefix) {
    return prefix.status();
  }
  auto parallel_upload = PrepareParallelUpload(client, bucket, "my_very_large_object", 16,
      prefix->name());
  if (!parallel_upload) {
    return parallel_upload.status();
  }
  std::vector<std::thread> uploaders;
  for (std::size_t i = 0; i < 16; ++i) {
    stream = parallel_upload->Shard(i);
    if (!stream) {
      return stream.status();
    }
    uploaders.emplace_back([i, stream]  { /* write to the stream */ });
  }
  for (auto &uploader : uploaders) {
    uploader.join();
  }
  // all streams completed, should return immediately
  auto object = parallel.PollForCompletion(std::chrono::seconds(0));
  Status cleanup_status = parallel.Cleanup();
```

If the upload is interrupted at any point, the only difference is that the user should pass a `RestoreParallelUploadSession` to `PrepareParallelUpload`. Individual streams need to be checked for last offset, just like regular resumable streams.

There is nothing which prevents the user to use `RestoreParallelUploadSession` on multiple machines and write to different streams on them.

# Under the hood

`PrepareParallelUpload()` simply creates an object with a random name called the *state*. The state contains:
* the information needed to create the final object
* the information about every individual stream (resumable upload session Id and whether it succeeded, is in progress or failed)

The `resumable_session_id` for the `ParallelUploadState` is basically the name of the object holding the state.

Below is the description of how the `ParallelUploadState`'s member functions are implemented
```C++
struct ParallelUploadState {
  // This is basically the path to the object holding the state.
  std::string resumable_session_id() const;

  // This is just an accessor to a field in the state.
  std::size_t NumShards() const;

  // If the state contains a `resumable_session_id` for such an index, return a stream for it, otherwise, 
  // create a resumable session, upload the altered state and return the relevant stream. The stream
  // has to have Close() overridden, such that once the stream is successfully closed, this fact is 
  // reflected in the state. If this was the last stream to finish, also invoke ComposeMany (and store its 
  // result in the state) before returning to the user that the stream is successfully closed.
  StatusOr<GcsObjectWriteStream> Shard(std::size_t idx);

  // Periodically reread the state to check if the operation has finished.
  template< class Rep, class Period >
  StatusOr<ObjectMetadata> PollForCompletion(
    const std::chrono::duration<Rep, Period>& rel_time);

  // Remove all the temporary files and the state itself.
  Status Cleanup();
};
```

Once this is written, we can wrap it into simpler functions, allowing e.g. single large file upload.

**Comment by coryan** [comment]
> Some additional (nice-to-have) requirements I thought of:
> * allow for separate processes writing to streams
> * a process restart/crash shouldn't imply starting from scratch

> Thoughts?

I think this is related to the vague ideas in #559, but if it is, this is more concrete and better.

Another thought is that of those two capabilities the second (recover from restart/crash) seems very useful (large uploads from preemptable VMs for example will suffer from interruptions).  The second seems less useful to me.  Uploading a file available to a single machine with multiple threads or multiple processes should yield more or less the same performance I would think. If the data is available to multiple machines, then maybe, but that feels like a rare use-case, or something that maybe [STS](https://cloud.google.com/storage-transfer/docs/) is a better match?

Finally, I would like us to implement the single process version first.

**Comment by dopiera** [comment]
FWIW, a quick test of uploading 1GiB file via parallel uploads implemented in #3279 resulted in almost 4x speed up - from 30s to 8s.

**Comment by dopiera** [comment]
Machine: `n2-standard-16` in `europe-west1-b`
GCS bucket in `europe-west1`
CPU usage never exceeded 20%
Memory (RSS) high watermark of the benchmark was 4.5 GB (machine has 64GB)
The file was placed in `/dev/shm`

```
Results for file_size=33554432 num_shards=1 avg=1035ms stddev=71ms min=929ms max=1144ms avg_bw=30MB/s num=8
Results for file_size=33554432 num_shards=4 avg=969ms stddev=81ms min=880ms max=1107ms avg_bw=33MB/s num=8
Results for file_size=33554432 num_shards=16 avg=2264ms stddev=147ms min=2062ms max=2536ms avg_bw=14MB/s num=8
Results for file_size=33554432 num_shards=64 avg=7513ms stddev=631ms min=6413ms max=8410ms avg_bw=4MB/s num=8
Results for file_size=33554432 num_shards=128 avg=13435ms stddev=339ms min=12897ms max=14044ms avg_bw=2MB/s num=8
Results for file_size=134217728 num_shards=1 avg=3867ms stddev=281ms min=3394ms max=4219ms avg_bw=33MB/s num=8
Results for file_size=134217728 num_shards=4 avg=1719ms stddev=191ms min=1522ms max=2045ms avg_bw=74MB/s num=8
Results for file_size=134217728 num_shards=16 avg=2148ms stddev=180ms min=1972ms max=2510ms avg_bw=59MB/s num=8
Results for file_size=134217728 num_shards=64 avg=7129ms stddev=336ms min=6556ms max=7760ms avg_bw=17MB/s num=8
Results for file_size=134217728 num_shards=128 avg=13556ms stddev=286ms min=13243ms max=14059ms avg_bw=9MB/s num=8
Results for file_size=1073741824 num_shards=1 avg=29291ms stddev=1068ms min=28045ms max=31372ms avg_bw=34MB/s num=8
Results for file_size=1073741824 num_shards=4 avg=8503ms stddev=136ms min=8349ms max=8818ms avg_bw=120MB/s num=8
Results for file_size=1073741824 num_shards=16 avg=4313ms stddev=352ms min=3913ms max=4838ms avg_bw=237MB/s num=8
Results for file_size=1073741824 num_shards=64 avg=8241ms stddev=522ms min=7765ms max=9354ms avg_bw=124MB/s num=8
Results for file_size=1073741824 num_shards=128 avg=14179ms stddev=334ms min=13752ms max=14760ms avg_bw=72MB/s num=8
Results for file_size=4294967296 num_shards=1 avg=118859ms stddev=1968ms min=116353ms max=122359ms avg_bw=34MB/s num=8
Results for file_size=4294967296 num_shards=4 avg=33075ms stddev=608ms min=32571ms max=34011ms avg_bw=123MB/s num=8
Results for file_size=4294967296 num_shards=16 avg=10802ms stddev=285ms min=10409ms max=11280ms avg_bw=379MB/s num=8
Results for file_size=4294967296 num_shards=64 avg=10087ms stddev=411ms min=9328ms max=10789ms avg_bw=406MB/s num=8
Results for file_size=4294967296 num_shards=128 avg=15821ms stddev=353ms min=15356ms max=16540ms avg_bw=258MB/s num=8
```

It looks like a good strategy is to split the file into shards which are no less than 32MB but cap their number to 64.

**Comment by coryan** [comment]
> It looks like a good strategy is to split the file into shards which are no less than 32MB but cap their number to 64.

I expext that the cap on the number of shards depends on the number of cores: GCE instances get more bandwidth as you add cores.  Presumably on a VM with 16 cores you could create more shards (if the file is large enough to justify it).

**Comment by dopiera** [comment]
> > It looks like a good strategy is to split the file into shards which are no less than 32MB but cap their number to 64.
> 
> I expext that the cap on the number of shards depends on the number of cores: GCE instances get more bandwidth as you add cores. Presumably on a VM with 16 cores you could create more shards (if the file is large enough to justify it).

To be perfectly honest, I don't understand these results. The things that I thought were true at first:
- each stream gives us roughly 30MB/s provided that it's long enough (around 30MB/s)

**Comment by dopiera** [comment]
> I expect that the cap on the number of shards depends on the number of cores: GCE instances get more bandwidth as you add cores. Presumably on a VM with 16 cores you could create more shards (if the file is large enough to justify it).

To be perfectly honest, I don't understand the reason for these results. My guess is that this is what happens:
- we can write at most 30MB/s from one stream
- the stream needs to be large enough to get to this 30MB/s
- the more streams there are, the larger the streams need to be to get to 30MB/s (the largest example shows that 256MB is not enough with 128 streams, but it perfectly enough when there are only 4 streams)

I didn't want to over-engineer the testing this time, so I figured that:
- with 64 streams we can do around 10Gib/s
- all the machine types (except `n1-standard-1`) have at least 10Gib network
- the only storage option which can go significantly faster than this is local SSD

This means that for a scenario of uploading a single file, we'd not have sane deafults for `n1-standard-1` (which would probably be short of CPU and memory to use parallel upload in the first place). If the file is placed on a local SSD we'd only use 50% of the available bandwidth, which is not too bad either, IMO. I intentionally neglected scenarios when the file is on a RAID0 spanning multiple devices, etc.

If this doesn't suit the user's needs, the user might force whatever number they wish.

**Comment by coryan** [comment]
> * all the machine types (except `n1-standard-1`) have at least 10Gib network

nit: I am not sure that is right.  I think you get 2Gpbs per core, up to 32Gpbs with 16 cores:

https://cloud.google.com/blog/products/gcp/5-steps-to-better-gcp-network-performance?hl=ml

**Comment by dopiera** [comment]
> > * all the machine types (except `n1-standard-1`) have at least 10Gib network
> 
> nit: I am not sure that is right. I think you get 2Gpbs per core, up to 32Gpbs with 16 cores:
> 
> https://cloud.google.com/blog/products/gcp/5-steps-to-better-gcp-network-performance?hl=ml

This is where I took it from:
https://cloud.google.com/compute/docs/machine-types

**Comment by coryan** [comment]
@dopiera can we close this issue now?

**Comment by dopiera** [comment]
I think after the next PR (which is ready module a bug which I'm still debugging now). The next PR will make `CreateUploadShards` and `ParallelUploadFile` resumable.

**Comment by dopiera** [comment]
FWIW, the stateful version (#3389) got even to 1.55 GiB/s once. These are some quick results:
```sh
$ GOOGLE_APPLICATION_CREDENTIALS=/home/marek/my_service_key.json /usr/bin/time -v ./google/cloud/storage/benchmarks/storage_parallel_uploads_benchmark --directory=/dev/shm/asd  --thread-count=1 --minimum-object-size=$(( 32 * 1024 * 1024 * 1024 )) --maximum-object-size=$(( 32 * 1024 * 1024 * 1024 )) --maximum-sample-count=10 --bucket-prefix=paralleluploads --region=EUROPE-WEST1 --project-id=grass-clump-479 --minimum-num-shards=64 --maximum-num-shards=64
# Running test on bucket: paralleluploadsf4069hwbme9t45dqrk6vabjm6f5kmorzg46t02orxez5j6ap
# Start time: 2020-02-18T13:32:10.815591143Z
# Region: EUROPE-WEST1
# Duration: 900s
# Thread Count: 1
# Min Object Size: 34359738368
# Max Object Size: 34359738368
# Min Object Size (MiB): 32768
# Max Object Size (MiB): 32768
# Min Number of Shards: 64
# Max Number of Shards: 64
# Build info: v1.11.0;GNU 7.4.0; 
FileSize,ShardCount,UploadTimeMs
34359738368,64,25207
34359738368,64,21940
34359738368,64,21988
34359738368,64,21845
34359738368,64,21723
34359738368,64,22344
34359738368,64,20513
34359738368,64,21882
34359738368,64,22173
34359738368,64,22559
        Command being timed: "./google/cloud/storage/benchmarks/storage_parallel_uploads_benchmark --directory=/dev/shm/asd --thread-count=1 --minimum-object-size=34359738368 --maximum-object-size=34359738368 --maximum-sample-count=10 --bucket-prefix=paralleluploads --region=EUROPE-WEST1 --project-id=grass-clump-479 --minimum-num-shards=64 --maximum-num-shards=64"
        User time (seconds): 1228.84
        System time (seconds): 521.13
        Percent of CPU this job got: 487%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 5:58.70
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 2242780
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 17798189
        Voluntary context switches: 2165472
        Involuntary context switches: 82640
        Swaps: 0
        File system inputs: 0
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```
